tests/test_layers.py

```diff
@@ -63,3 +68,35 @@ def test_override_settings(self):
     message = {"type": "test.message"}
     await layer.send("test.channel", message)
     assert message == await layer.receive("test.channel")
+
+
+@pytest.mark.parametrize(
+    "group_name,expected_valid",
+    [("¯\_(ツ)_/¯", False), ("chat", True), ("chat" * 100, False)],
+)
+def test_valid_group_name(group_name, expected_valid):
+    layer = InMemoryChannelLayer()
+    if expected_valid:
+        layer.valid_group_name(group_name)
+    else:
+        with pytest.raises(TypeError) as e:
+            layer.valid_group_name(group_name)
+            assert e.value.message.startswith("Group")
+            assert group_name in e.value.message
+            assert "< {}".format(MAX_NAME_LENGTH) in e.value.message
+
+
+@pytest.mark.parametrize(
+    "channel_name,expected_valid",
+    [("¯\_(ツ)_/¯", False), ("chat", True), ("chat" * 100, False)],
+)
+def test_valid_channel_name(channel_name, expected_valid):
+    layer = InMemoryChannelLayer()
+    if expected_valid:
+        layer.valid_channel_name(channel_name)
+    else:
+        with pytest.raises(TypeError) as e:
+            layer.valid_channel_name(channel_name)
+            assert e.value.message.startswith("Channel")
+            assert channel_name in e.value.message
+            assert "< {}".format(MAX_NAME_LENGTH) in e.value.message
```

Comment:
Is it possible to refactor the two test functions please ? It really looks like copy/paste, when only the callback changes from one test to the other.

Reply:
Instead of

```

        with pytest.raises(TypeError) as e:
            layer.valid_group_name(group_name)
            assert e.value.message.startswith("Group")
            assert group_name in e.value.message
            assert "< {}".format(MAX_NAME_LENGTH) in e.value.message

        with pytest.raises(TypeError) as e:
            layer.valid_channel_name(channel_name)
            assert e.value.message.startswith("Channel")
            assert channel_name in e.value.message
            assert "< {}".format(MAX_NAME_LENGTH) in e.value.message

```

Try something like

```
def assert_valid_name(kind, cb, name):
        with pytest.raises(TypeError) as e:
            cb(name)

        assert e.value.message.startswith(kind.upper())
        assert name in e.value.message
        assert "< {}".format(MAX_NAME_LENGTH) in e.value.message

```

Note that this example also minimizes the code inside TypeError expect raise block.

Reply:
Actually I'm not convinced by the whole approach to this test code.

Just test that:

- [ ] valid_channel_name passes on valid name
- [ ] valid_group_name passes on valid name
- [ ] valid_channel_name raises TypeError on invalid name
- [ ] valid_group_name raises TypeError on invalid name

You don't need to test that Python .format() works.

Reply:
I agree it's not required to test format().

I'll use pytest's `parametrize` to generate test cases for both methods.